### PR TITLE
[NFC] Move runtime API call generator into Optimizer/Runtime

### PR DESCRIPTION
**Pull request by jeanPerier**
In order to be able to emit runtime calls in FIR transformation passes,
move the runtime call generators into Optimizer/Runtime.
RTBuilder.h is moved there, as well as DerivedRuntime.h, NumericRuntime.h,
ReductionRuntime.h, CharacterRuntime.h, and TransformationalRuntime.h

All the methods are moved from the Fortran::lower namespace into a new
fir::runtime namespace.

addingIntrinsics.md is updated to reflect the namespace and folder change.

Runtime.[h, cpp] is not moved yet because it uses parse tree nodes.
There is no urgent need to move it now, so I did not attempted to split
it in this PR.

Allocatable.cpp and IO.cpp make direct usages of RTBuilder.h, it may be
cleaner to handle the runtime call generation details in new
Optimizer/xxx files, but this is not done here since this calls are not
currently needed in lib/Optimizer.

Tested OK with both static and shared library builds.

### CODE REVIEW

File: flang/include/flang/Optimizer/Runtime/RTBuilder.h [hunk 1]
```diff
@@ -14,28 +14,29 @@
 ///
 //===----------------------------------------------------------------------===//
 
-#ifndef FORTRAN_LOWER_RTBUILDER_H
-#define FORTRAN_LOWER_RTBUILDER_H
+#ifndef FORTRAN_OPTIMIZER_RUNTIME_RTBUILDER_H
+#define FORTRAN_OPTIMIZER_RUNTIME_RTBUILDER_H
 
-#include "flang/Lower/ConvertType.h"
+#include "flang/Common/Fortran.h"
```
Comment (schweitzpgi) [inline flang/include/flang/Optimizer/Runtime/RTBuilder.h:20]: Is this a new dependency? Optimizer -> Common?

File: flang/include/flang/Optimizer/Runtime/RTBuilder.h [hunk 2]
```diff
@@ -14,28 +14,29 @@
...
 // and std::complex are layout compatible, but not compatible in all ABI call
 // interface (e.g. X86 32 bits). _Complex is not standard C++, so do not use
 // it here.
 struct c_float_complex_t;
 struct c_double_complex_t;
 
-namespace Fortran::lower {
+namespace Fortran::runtime {
+class Descriptor;
+}
```
Comment (schweitzpgi) [inline flang/include/flang/Optimizer/Runtime/RTBuilder.h:37]: Should we consider an abstraction layer sometime soon? We might want to at least consider what a move of runtime interfaces to `.td` files vs. using reflection on a set of C++ header files would look like.

File: flang/lib/Optimizer/Runtime/Transformational.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
+//===-- Transformational.cpp ------------------------------------*- C++ -*-===//
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Runtime/Transformational.cpp:1]: I wonder why git failed to twig this one was moved?

**Review by schweitzpgi — APPROVED** [review]
+1

File: flang/lib/Optimizer/Runtime/Transformational.cpp (hunk 3 above)
Comment (jeanPerier) [reply flang/lib/Optimizer/Runtime/Transformational.cpp:1]: I do not know either, I had used git mv. I have improved the PR history by splitting the PR in a commit that renames the files without any changes (528737c96dbcf37b1f5de14d3d3a2945f822e008), and another one with the content changes (187b675fad1356cb3846653701cb29b0a5d4a845), but the overall GitHub PR diff still does not show this has a rename.

**Pull request merged**